### PR TITLE
test(bob-escrow): fork and fuzz

### PR DESCRIPTION
**Pull request by andreivladbrg**

### CODE REVIEW

File: bob/tests/escrow/integration/fuzz/cancelOrder.t.sol [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        minBuyAmount = boundUint128(minBuyAmount, 1, MAX_UINT128);
+        expiryTime = boundUint40(expiryTime, getBlockTimestamp() + 1, getBlockTimestamp() + 5 * 365 days);
+        timeJump = boundUint40(timeJump, 0, uint40(10 * 365 days));
+
+        // Deal sell tokens to seller.
+        deal({ token: address(sellToken), to: users.seller, give: sellAmount });
+
+        // Create the order as seller.
+        setMsgSender(users.seller);
+        uint256 orderId = escrow.createOrder({
```
Comment (smol-ninja) [inline bob/tests/escrow/integration/fuzz/cancelOrder.t.sol]: I think this is an overkill. We should just test cancelling order based on timestamp warp.

```solidity
function testFuzz_CancelOrder(uint64 warpTime)
```

where we warp time to a random value between order creation to until 1 month after expiry. And the fuzz test should allow cancelling it.

File: bob/tests/escrow/integration/fuzz/createOrder.t.sol [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    function testFuzz_CreateOrder(uint128 sellAmount, uint128 minBuyAmount, uint40 expiryTime) external {
+        sellAmount = boundUint128(sellAmount, 1, MAX_UINT128);
+        minBuyAmount = boundUint128(minBuyAmount, 1, MAX_UINT128);
+
+        // If non-zero, bound to [now+1, now+5years]; otherwise keep 0 (never expires).
+        if (expiryTime != 0) {
+            expiryTime = boundUint40(expiryTime, getBlockTimestamp() + 1, getBlockTimestamp() + 5 * 365 days);
+        }
+
+        // Deal sell tokens to seller so any uint128 amount works.
```
Comment (smol-ninja) [inline bob/tests/escrow/integration/fuzz/createOrder.t.sol]: What does "so any uint128 amount works" mean?

File: bob/tests/escrow/integration/fuzz/fillOrder.t.sol [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+import { ISablierEscrow } from "src/interfaces/ISablierEscrow.sol";
+import { Errors } from "src/libraries/Errors.sol";
+import { Escrow } from "src/types/Escrow.sol";
+
+import { Integration_Test } from "./../Integration.t.sol";
+
+contract FillOrder_Integration_Fuzz_Test is Integration_Test {
+    /// @dev Designated buyer, default trade fee (1%).
+    function testFuzz_FillOrder(
```
Comment (smol-ninja) [inline bob/tests/escrow/integration/fuzz/fillOrder.t.sol]: It should also just have fuzzed timestamp and a fuzzed amount. The test should test that

- Any buyer can fill order until expiry with fuzzed amount >= min amount. So this test would have a fuzzed buyer, fuzzed amount and a fuzzed timestamp.
- If order has designated buyer, no buyer can fill the order. So that would have a fuzzed buyer address, a fuzzed timestamp and a fuzzed amount.

File: bob/tests/escrow/integration/fuzz/fillOrder.t.sol [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+            buyAmount: buyAmount,
+            expiryTime: expiryTime,
+            timeJump: timeJump,
+            buyer: users.buyer,
+            fee: DEFAULT_TRADE_FEE
+        });
+    }
+
+    /// @dev Designated buyer, zero trade fee — exercises the `if (currentTradeFee > 0)` skip branch.
+    function testFuzz_FillOrder_GivenTradeFeeZero(
```
Comment (smol-ninja) [inline bob/tests/escrow/integration/fuzz/fillOrder.t.sol]: Why do we need a test with trade fee as zero? Isn't a test with default trade fee but different buy amount enough?

File: bob/tests/escrow/integration/fuzz/fillOrder.t.sol [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+            buyAmount: buyAmount,
+            expiryTime: expiryTime,
+            timeJump: timeJump,
+            buyer: users.buyer,
+            fee: ZERO
+        });
+    }
+
+    /// @dev Open order (buyer = address(0)), default trade fee — exercises the open-order access path.
+    function testFuzz_FillOrder_GivenOpenOrder(
```
Comment (smol-ninja) [inline bob/tests/escrow/integration/fuzz/fillOrder.t.sol]: Open order is confusing here since there is a status called `OPEN`. We use the term no designated buyer for this

**Review by smol-ninja — CHANGES_REQUESTED** [review]

File: bob/tests/escrow/integration/fuzz/cancelOrder.t.sol (hunk 1 above)
Comment (andreivladbrg) [reply bob/tests/escrow/integration/fuzz/cancelOrder.t.sol]: your point is that we should use the default order?

File: bob/tests/escrow/integration/fuzz/createOrder.t.sol (hunk 2 above)
Comment (andreivladbrg) [reply bob/tests/escrow/integration/fuzz/createOrder.t.sol]: am not sure - claude added it xD

my guess is overflow 

we can remove it

File: bob/tests/escrow/integration/fuzz/cancelOrder.t.sol (hunk 1 above)
Comment (smol-ninja) [reply bob/tests/escrow/integration/fuzz/cancelOrder.t.sol]: Yes and fuzz timestamp only. Because for an open order, cancel should work always. Creating order with different value has no effect on it.

Comment (andreivladbrg) [reply bob/tests/escrow/integration/fuzz/cancelOrder.t.sol]: the reason i've created custom orders is that, unlike in flow where we have the [fixtures](https://github.com/sablier-labs/lockup/blob/staging/flow/tests/integration/fuzz/Fuzz.t.sol#L19) (defaults + custom ones), i thought having only new orders would be useful

wdyt?

Comment (smol-ninja) [reply bob/tests/escrow/integration/fuzz/cancelOrder.t.sol]: `cancel` only has one check for `OPEN` order: "revert if `msg.sender != order.seller`". It does not depend on other parameters of the order. 

Therefore, given our style for fuzz tests in other protocols, it only makes sense to have two tests (1) `testFuzz_ReverWhen_CallerNotSeller(address caller)` (2) `testFuzz_CancelOrder(uint64 warpTime)` and then just use default order.

[An example from Lockup](https://github.com/sablier-labs/lockup/blob/staging/lockup/tests/integration/fuzz/lockup/cancel.t.sol).

Comment (andreivladbrg) [reply bob/tests/escrow/integration/fuzz/cancelOrder.t.sol]: > `cancel` only has one check for `OPEN` order: "revert if `msg.sender != order.seller`". It does not depend on other parameters of the order.

ik ik. 

i don't have a strong preference - i used this approach because we have in solana as well - i tested the waters here xD

would u like to make the change?

File: bob/tests/escrow/integration/fuzz/fillOrder.t.sol (hunk 4 above)
Comment (andreivladbrg) [reply bob/tests/escrow/integration/fuzz/fillOrder.t.sol]: i added because of this check:

https://github.com/sablier-labs/lockup/blob/3051b8594b55f05652f8cdd757d8f2f873f40f42/bob/src/SablierEscrow.sol#L207

i think its useful to keep as there less operations done in comparison with `fee > 0`

File: bob/tests/escrow/integration/fuzz/fillOrder.t.sol (hunk 5 above)
Comment (andreivladbrg) [reply bob/tests/escrow/integration/fuzz/fillOrder.t.sol]: i am not sure i understand your point

that’s what it should test - `fillOrder` when the status is `OPEN`

Comment (smol-ninja) [reply bob/tests/escrow/integration/fuzz/fillOrder.t.sol]: `OPEN` status means order can be filled.  In the comment, open implies `buyer = address(0)`. Both have different meanings.

File: bob/tests/escrow/integration/fuzz/cancelOrder.t.sol (hunk 1 above)
Comment (smol-ninja) [reply bob/tests/escrow/integration/fuzz/cancelOrder.t.sol]: https://github.com/sablier-labs/lockup/pull/1419/changes/160dc6aa6ee1a2ea32a8c6a219e158f844493a1f

I also added status to some modifiers to make them explicit about it.

File: bob/tests/escrow/integration/fuzz/createOrder.t.sol (hunk 2 above)
Comment (smol-ninja) [reply bob/tests/escrow/integration/fuzz/createOrder.t.sol]: https://github.com/sablier-labs/lockup/pull/1419/changes/160dc6aa6ee1a2ea32a8c6a219e158f844493a1f

File: bob/tests/escrow/fork/Fork.t.sol [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+import { Base_Test } from "./../Base.t.sol";
+
+/// @notice Base logic needed by the fork tests.
+abstract contract Fork_Test is Base_Test {
+    IERC20 internal constant USDC = IERC20(0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48); // sell token
+    IERC20 internal constant WETH = IERC20(0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2); // buy token
+
+    function setUp() public virtual override {
+        // Fork Ethereum Mainnet at the latest block number.
+        vm.createSelectFork({ urlOrAlias: "ethereum" });
```
[the remark this answers is not in the document]
Comment (smol-ninja) [reply bob/tests/escrow/fork/Fork.t.sol:15]: Its OK to fork it at latest block. This helps us to test protocol against the latest fork.

File: bob/tests/escrow/integration/fuzz/fillOrder.t.sol (hunk 4 above)
Comment (smol-ninja) [reply bob/tests/escrow/integration/fuzz/fillOrder.t.sol]: OK sounds good. But we should refactor the test to simplify them a bit similar to cancel order. See [this comment](https://github.com/sablier-labs/lockup/pull/1419#discussion_r2864926192).

File: bob/tests/escrow/integration/fuzz/fillOrder.t.sol (hunk 5 above)
Comment (andreivladbrg) [reply bob/tests/escrow/integration/fuzz/fillOrder.t.sol]: right, iget it now

97b7ef0d21168bfc4ed6b6b524bd0786c846621b

File: bob/tests/escrow/integration/fuzz/fillOrder.t.sol (hunk 3 above)
Comment (andreivladbrg) [reply bob/tests/escrow/integration/fuzz/fillOrder.t.sol]: f3e9a42ab1f5dd411ffd5fd90136a734da5c7bd0

**Comment by andreivladbrg** [comment]
@smol-ninja does it look good now?

**Review by smol-ninja — APPROVED** [review]
LGTM.
